docs/development/how_to_add_a_route.md

```diff
@@ -0,0 +1,175 @@
+<!-- Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS -->
+
+The goal of this document is to cover all the necessary steps to add an API route to Parsec.
+
+In summary:
+1. Add the schema describing the new API route
+2. Generate binding code and serialization tests
+3. Implement server-side code (In-memory & PostgreSQL) and tests
+4. Implement client-side code and tests
+
+> Note that you will need a working development environment. See (./README.md).
+## Schema generation
+
+Let's add a new `dummy` route to `invited_cmds`.
+
+> API routes are also called "commands" in Parsec. There are three command families: `anonymous_cmds` (authentication not required), `authenticated_cmds` (authentication required) and `invited_cmds` (invitation required).
+
+In [this folder](../../misc/libparsec/crates/protocol/schema/invited_cmds/) add a `dummy.json5` schema describing the API route.
+
+```json5
+[
+    {
+        "major_versions": [
+            4
+        ],
+        "req": {
+            "cmd": "dummy",
+            "fields": [
+                {
+                    "name": "dummy_field",
+                    "type": "String"
+                }
+            ]
+        },
+        "reps": [
+            {
+                "status": "dummy_rep_with_fields",
+                "fields": [
+                    {
+                        "name": "dummy_rep_field",
+                        "type": "String"
+                    }
+                ]
+            },
+            {
+                "status": "dummy_rep_without_fields",
+            }
+        ]
+    }
+]
+
+```
+For a full description of the accepted format see [the parsing/generating script](../../misc/gen_protocol_typings.py).
+
+> Note that the files are .json5. This is to allow for comments inside these files. But the parser we are using only supports .json files and we're removing the comments ourselves. So don't use any other specific feature of the json5 format (trailing commas, I'm looking at you 👀).
+
+Then from the root directory, run the following scripts to generate binding code.
+
+```shell
+python make.py r
```

Comment:
This is the wrong command, it should be `python misc/gen_protocol_typings.py`

Reply:
Ok so first we should run gen_protocol_typing and then make.py r ?

Reply:
they are orthogonal:

- `gen_protocol_typing.py` regenerates some python files that gets commited to the project
- `make.py r` (re)build `server/parsec/_parsec.cpython-312-x86_64-linux-gnu.so`

Modifying Python file doesn't invalidate `server/parsec/_parsec.cpython-312-x86_64-linux-gnu.so` so there is no need to run `make.py r` after `gen_protocol_typing.py`.

What is misleading is the reason for running `gen_protocol_typing.py` is because the protocol schema have been updated... which is also a reason to run `make.py r` ! (so you could say `make.py r` and `gen_protocol_typing.py` are cousin as they are triggered by the same reason, but don't have a parenting relationship between them.

Reply:
Ok so if stuff is changed everywhere, we need to run both in either order ?

Reply:
> Ok so if stuff is changed everywhere, we need to run both in either order ?

yup ! ^^